cellrank/tl/kernels/_velocity_schemes.py

```diff
@@ -0,0 +1,345 @@
+    DOT_PRODUCT = "dot_product"
+    COSINE = "cosine"
+    CORRELATION = "correlation"
+
+
+try:
+    import jax.numpy as jnp
+    from jax import jit, hessian
+
+    @jit
+    def _softmax_jax(x: np.ndarray, softmax_scale) -> np.ndarray:
+        numerator = x * softmax_scale
+        numerator = jnp.exp(numerator - jnp.max(numerator))
+        return numerator / jnp.sum(numerator)
+
+    @jit
+    def _softmax_masked_jax(
+        x: np.ndarray, mask: np.ndarray, softmax_scale
+    ) -> np.ndarray:
+        numerator = x * softmax_scale
+        numerator = jnp.exp(numerator - jnp.nanmax(numerator))
+        numerator = jnp.where(mask, 0, numerator)  # essential
+
+        return numerator / jnp.nansum(numerator)
+
+    @partial(jit, static_argnums=(3, 4))
+    def _predict_transition_probabilities_jax(
+        X: np.ndarray,
+        W: np.ndarray,
+        softmax_scale: float = 1.0,
+        center_mean: bool = True,
+        scale_by_norm: bool = True,
+    ):
+        if center_mean:
+            # pearson correlation, otherwise cosine
+            W -= W.mean(axis=1)[:, None]
+            X -= X.mean()
+
+        if scale_by_norm:
+            denom = jnp.linalg.norm(X) * jnp.linalg.norm(W, axis=1)
+            mask = jnp.isclose(denom, 0)
+            denom = jnp.where(jnp.isclose(denom, 0), 1, denom)  # essential
+            return _softmax_masked_jax(W.dot(X) / denom, mask, softmax_scale)
+
+        return _softmax_jax(W.dot(X), softmax_scale)
+
+    _predict_transition_probabilities_jax_H = hessian(
+        _predict_transition_probabilities_jax
+    )
+
+    _HAS_JAX = True
+except ImportError:
+    _HAS_JAX = False
+
+    hessian = jit = lambda _: _
+
+    def _predict_transition_probabilities_jax(*_args, **_kwargs):
+        raise NotImplementedError("No `jax` installation found.")
+
+    _predict_transition_probabilities_jax_H = _predict_transition_probabilities_jax
+
+
+@njit(**jit_kwargs)
+def _softmax(x: np.ndarray, softmax_scale: float) -> Tuple[np.ndarray, np.ndarray]:
+    numerator = x * softmax_scale
+    numerator = np.exp(numerator - np.max(numerator))
+    return numerator / np.sum(numerator), x
+
+
+@njit(**jit_kwargs)
+def _softmax_masked(
+    x: np.ndarray, mask: np.ndarray, softmax_scale: float
+) -> Tuple[np.ndarray, np.ndarray]:
+    numerator = x * softmax_scale
+    numerator = np.exp(numerator - np.nanmax(numerator))
+    numerator = np.where(mask, 0, numerator)  # essential
+
+    return numerator / np.nansum(numerator), x
+
+
+@njit(parallel=False, **jit_kwargs)
+def _predict_transition_probabilities_numpy(
+    X: np.ndarray,
+    W: np.ndarray,
+    softmax_scale: float = 1.0,
+    center_mean: bool = True,
+    scale_by_norm: bool = True,
+) -> Tuple[np.ndarray, np.ndarray]:
+    if center_mean:
+        # pearson correlation
+        W -= np.expand_dims(np_mean(W, axis=1), axis=1)
+
+    if X.shape[0] == 1:
+        if center_mean:
+            # pearson correlation
+            X = X - np.mean(X)
```

Comment:
Why is this different here than in the Jax case?

Reply:
Is that because this works for the backwards case as well?

Reply:
If you refer to the `np_mean`, it's because in numba, `axis` cannot be specified in `np.mean`.
If you refer to the `X = X - ...` (instead of `X -= ...)`, it's because I can't use inplace assignment there.

Reply:
Ah not I was thinking of why we have the extra if statement checking for `X.shape[0] == 1`, and I do assume that's because this function can handle the backward case as well?